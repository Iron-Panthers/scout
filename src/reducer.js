export const reducer = (state, action) => {
  switch (action.type) {
    case "next_mode":
      const modes = [ "Configure", "Scout", "Review", "Scan" ]
      return {
        ...state,
        mode: modes[modes.indexOf(state.mode) + 1]
      }
    // base reducer, no special behavior
    case "set":
      console.log(action.val)
      return {
        ...state,
        [action.prop]: action.val
      }
    default:
      return state
  }
}

const shooting = {
  innerOuterSucc: 0,
  innerOuterFail: 0,
  lowerSucc: 0,
  outerSucc: 0
}

export const initialState = {
  mode: "Configure",
  team: undefined,
  matchType: "Test",
  matchNum: undefined,
  timeLeft: 150, //seconds
  phase: "auto", //auto, teleop, endgame
  auto: {
    pathType: undefined,
    ...shooting
  },
  tele: {
    ...shooting
  },
  end: {
    climb: false,
    park: false,
    level: false, //level vs not level, not level is false
    levelTime: undefined //value of timeLeft when level is set
  },
  underTrench: false,
  defense: false,
  problems: false,
  comments: "",
}