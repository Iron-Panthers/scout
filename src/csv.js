import { unparse, parse } from "papaparse"
import { initialState } from "./reducer"

/** remove unwanted keys from state */
export const filterState = ({ mode, phase, undoStack, ...state }) => state

/** flattens, cleans a state object such that its keys are hyphen separated of their parents*/
export const cleanState = (rawState) =>
  Object.entries(filterState(rawState)).reduce((obj, [key, val]) => {
    if (typeof val === "object") {
      Object.entries(val).forEach(([key2, val2]) => {
        obj[`${key}-${key2}`] = val2
      })
    } else {
      obj[key] = val
    }
    return obj
  }, {})

/** the fields on a state - the flattened keys */
export const fields = Object.entries(cleanState(initialState)).map(
  ([key]) => key
)

/** the csv header for a csv ed state */
export const header = unparse({ fields })

/** converts state into a csv string */
export const stateToCsv = (state) =>
  unparse(
    {
      fields,
      data: Object.entries(cleanState(state)).map(([, val]) => val),
    },
    { header: false }
  )

/** converts the body of a csv string into a flattened state object */
export const parseCsvBody = (body) => {
  const obj = parse(`${header}\r\n${body}`)
  const headArray = obj.data[0]
  const bodyArray = obj.data[1]

  let returnObj = {}

  headArray.forEach((row, i) => {
    const value = bodyArray[i]
    returnObj[row] = value
  })

  return returnObj
}

export const castType = (string) => {
  if (Number.isFinite(string)) return Number.parseFloat(string)
  if ((string === "true") | (string === "false"))
    return string === "true" ? true : false
  return string
}

/** shallow iteration over an object, to cast types into their respective values if applicable or leave them as strings */
export const castTypes = (object) => {
  return Object.entries(object).reduce((obj, [key, value]) => {
    return (obj[key] = castType(value))
  }, {})
}
