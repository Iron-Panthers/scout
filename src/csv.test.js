import { filterState } from "./csv"
import { initialState } from "./reducer"

it("matches snapshot for filter on initialState", () => {
  expect(JSON.stringify(filterState(initialState))).toMatchSnapshot()
})
