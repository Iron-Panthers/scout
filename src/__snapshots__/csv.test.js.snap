// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object parsers and encoders matches snapshot for cleanState on initialState 1`] = `"{\\"version\\":1,\\"matchType\\":\\"Qualification\\",\\"auto-pickup\\":0,\\"auto-taxi\\":false,\\"auto-upperSucc\\":0,\\"auto-upperFail\\":0,\\"auto-lowerSucc\\":0,\\"auto-lowerFail\\":0,\\"teleop-upperSucc\\":0,\\"teleop-upperFail\\":0,\\"teleop-lowerSucc\\":0,\\"teleop-lowerFail\\":0,\\"endgame-climb\\":false,\\"endgame-fail\\":false,\\"endgame-level\\":\\"None\\",\\"wrongCargo\\":false,\\"defense\\":false,\\"scoutProblems\\":false,\\"robotProblems\\":false,\\"comments\\":\\"\\"}"`;

exports[`object parsers and encoders matches snapshot for filter on initialState 1`] = `"{\\"version\\":1,\\"matchType\\":\\"Qualification\\",\\"auto\\":{\\"pickup\\":0,\\"taxi\\":false,\\"upperSucc\\":0,\\"upperFail\\":0,\\"lowerSucc\\":0,\\"lowerFail\\":0},\\"teleop\\":{\\"upperSucc\\":0,\\"upperFail\\":0,\\"lowerSucc\\":0,\\"lowerFail\\":0},\\"endgame\\":{\\"climb\\":false,\\"fail\\":false,\\"level\\":\\"None\\"},\\"wrongCargo\\":false,\\"defense\\":false,\\"scoutProblems\\":false,\\"robotProblems\\":false,\\"comments\\":\\"\\"}"`;

exports[`object parsers and encoders matches snapshot for header 1`] = `"version,team,matchType,matchNum,auto-pickup,auto-taxi,auto-upperSucc,auto-upperFail,auto-lowerSucc,auto-lowerFail,teleop-upperSucc,teleop-upperFail,teleop-lowerSucc,teleop-lowerFail,endgame-climb,endgame-fail,endgame-level,wrongCargo,defense,scoutProblems,robotProblems,comments"`;

exports[`object parsers and encoders matches snapshot for state to csv on initialState 1`] = `"\\"1,,Qualification,,0,false,0,0,0,0,0,0,0,0,false,false,None,false,false,false,false,\\""`;
